compiler/rustc_middle/src/ty/typeck_results.rs

```diff
@@ -183,7 +183,8 @@ pub struct TypeckResults<'tcx> {
     /// we never capture `t`. This becomes an issue when we build MIR as we require
     /// information on `t` in order to create place `t.0` and `t.1`. We can solve this
     /// issue by fake reading `t`.
-    pub closure_fake_reads: FxHashMap<LocalDefId, Vec<(HirPlace<'tcx>, FakeReadCause, hir::HirId)>>,
+    pub closure_fake_reads:
+        FxIndexMap<LocalDefId, Vec<(HirPlace<'tcx>, FakeReadCause, hir::HirId)>>,
```

Comment:
Can this be a `LocalDefIdMap`? The order shouldn't matter.
Likewise for the others in `typeck_results`.

Reply:
Yes, that's what I would prefer too. However, it turned out to be not straightforward because for many of these maps, the `writeback` module will read the fcx-specific results, perform some transformation, and then write the transformed data back into `WritebackCx::typeck_results`. That can be done like [here](https://github.com/rust-lang/rust/blob/9227ff28aff55b252314076fcf21c9a66f10ac1e/compiler/rustc_hir_typeck/src/writeback.rs#L518-L535), except that most of the transformations involve calling `WritebackCx::resolve` which takes a `&mut` reference to `WritebackCx` and transitively to its contained `WritebackCx::typeck_results`. Because of that we cannot use it inside of `UnordItems::map` which [does not allow `FnMut`](https://github.com/rust-lang/rust/blob/9227ff28aff55b252314076fcf21c9a66f10ac1e/compiler/rustc_data_structures/src/unord.rs#L23-L33).

The alternatives I tried are:
- Wrapping `TypeckResults::tainted_by_errors` inside a `Lock`, so that `WritebackCx::resolve` does not need `&mut` references. But I really didn't like adding interior mutability to something that's a query result. It seemed like a potentially source of problems (or at least confusion).
- Extract the fcx-specific data into a sorted collection before iterating/transforming them (like is done [here](https://github.com/rust-lang/rust/blob/9227ff28aff55b252314076fcf21c9a66f10ac1e/compiler/rustc_hir_typeck/src/writeback.rs#L461). But that resulted in lots of boilerplate code and potentially an unnecessary performance hit. 

But I agree that it would be better to use `LocalDefIdMap`. Some further things that could be tried:
1. Come up with some helper functions that reduce the amount of boilerplate code needed for doing the sorting approach.
2. Refactor `writeback` and `TypeckResult` so that `UnordItems::map` can be called with an `Fn` instead of an `FnMut`. I'm not sure that's worth the trouble unless it would also improve the overall quality of the code in question.
3. Come up with a safe API in `unord` that allows to do the data mapping while still guaranteeing that we don't leak ordering information. That might be possible, but I'm not sure the resulting API would be easy to understand.
4. Add something like `UnordItems::map_unchecked(self, f: FnMut())` that puts the burden of making sure no ordering information is leaked on the programmer. I'm not sure I want to do that. It seems like a potential source of problems once the code in question is modified a few times by different people.

I'd be happy to try how far I get with option (1) or (3). I'm also interested in other alternatives to solve the problem.